README.md

[![Flutter Community: iframe_cashpay_plugin](https://fluttercommunity.dev/_github/header/iframe_cashpay_plugin)](https://github.com/fluttercommunity/community)


# iframe_cashpay_plugin

[![pub package](https://img.shields.io/pub/v/iframe_cashpay_plugin.svg)](https://pub.dartlang.org/packages/flutter_webview_plugin)

iframe_cashpay_plugin.
A plugin to add payments iframe_cashpay to your Flutter application.

![alt iframe_cashpay_plugin](https://raw.githubusercontent.com/Mazen-Aljaradi/iframe_cashpay_plugin/main/documentation/en_US/cashb.png)

## Platform Support

| Android |
|:-------:|
|   IOS   |
| :-----: |

## Getting Started

Before you start, create an APIs with the payment providers and follow the setup instructions:
https://documenter.getpostman.com/view/29850098/2s9YXcd5Fd

## Usage

To start using this plugin, add `iframe_cashpay_plugin` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/):

```yaml
dependencies:
  iframe_cashpay_plugin: ^2.0.4
```

### Example

```dart
import 'package:iframe_cashpay_plugin/iframe_cashpay_plugin.dart';
import 'package:iframe_cashpay_plugin/CashPayButton.dart';

class PaySampleAppState extends State<PaySampleApp> {
  
  Future<String> sendItems(itemList) async {
    //Send itemList for yor server and post CreateOrder.
    //iframeURL returned from Response CreateOrder
    //Documentation https://documenter.getpostman.com/view/17550185/2s93XzwN9o
    //String iframeURL = "**********************************************************";
    return iframeURL;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        backgroundColor: Colors.white,
        body: ListView(children: [
          CashPayButton(width:MediaQuery.of(context).size.width ,onTap: () async {
            await sendItems({
              {
                "itemName": "كتاب",
                "amount": 2000,
              },
              {
                "itemName": "ساعة",
                "amount": 5000,
              }
            }).then((iframeURL) => showModalBottomSheet(
                context: context,
                isDismissible: false,
                shape: const RoundedRectangleBorder(
                  borderRadius: BorderRadius.vertical(
                    top: Radius.circular(30.0),
                  ),
                ),
                isScrollControlled: true,
                builder: (BuildContext context) {
                  return Container(
                      padding: const EdgeInsets.only(
                          left: 10.0, right: 10.0, top: 10.0),
                      height: MediaQuery.of(context).size.height * 0.7,
                      //IframeCashPay widget displays the Cash E-wallet payment iframe.
                      /// This widget IframeCashPay displays the Cash E-wallet payment iframe.
                      ///
                      /// @param iframeURL .
                      /// @param onConfirmPayment.
                      /// @param onCancel.
                      /// @param onError.
                      /// @return message onConfirmPayment or onCancel or onError.
                      ///
                      /// Example:
                      ///
                      /// ```
                      ///  IframeCashPay( iframeURL: "https://########",
                      ///    onConfirmPayment: onConfirmPayment,
                      ///    onCancel: onCancel,
                      ///    onError: onError,);
                      /// ```
                      child: IframeCashPay(
                        iframeURL: iframeURL,
                        onConfirmPayment: onConfirmPayment,
                        onCancel: onCancel,
                        onError: onError,
                      ));
                }));})
        ]));
  }

  //Await for iFrameCashPay
  onConfirmPayment(message) {
    Navigator.pop(context);
    //After Confirmatin from iFrameCashPay.
    //Here use CheckOrderStatus to check order status.
    //Documentation https://documenter.getpostman.com/view/17550185/2s93XzwN9o
  }

//Await for iFrameCashPay
  onCancel(message) {
    //After Cancel from iFrameCashPay.
    Navigator.pop(context);
  }

//Await for iFrameCashPay
  onError(message) {
    //After return Error from iFrameCashPay.
    Navigator.pop(context);
  }
}
```


CHANGELOG.md
## [2.0.5]

Close.
## [2.0.4]

Improve code.

## [2.0.3]

Improve code.
## [2.0.2]

Improve code.
## [2.0.1]

Improve code.
## [2.0.0]

Improve code.
## [1.0.8]

Improve code.

## [1.0.7]

Improve docs.

## [1.0.6]

Improve docs.

## [1.0.5]

Improve docs.

## [1.0.4]

Improve docs.

## [1.0.3]

Improve docs.

## [1.0.2]

Improve docs.

## [1.0.1]

Improve docs.

## [0.0.9]

Improve docs.

## [0.0.8]

Improve docs.

## [0.0.7]

Improve docs.

## [0.0.6]

Fix pub dev score.

## [0.0.5]

Improve docs.


lib/CashPayButton.dart

import 'package:flutter/material.dart';

class CashPayButton extends StatelessWidget {


  CashPayButton({
    Key? key,

    this.width,

     this.alignment, this.margin, this.height, this.padding, this.decoration, this.child, this.onTap
  ,
  }) : super(
    key: key,
  );

  final Alignment? alignment;

  final EdgeInsetsGeometry? margin;

  final double? height;

  final double? width;

  final EdgeInsetsGeometry? padding;

  final BoxDecoration? decoration;

  final Widget? child;

  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    return alignment != null
        ? Align(
      alignment: alignment ?? Alignment.center,
      child: cashPayButtonWidget,
    )
        : cashPayButtonWidget;
  }

  Widget get cashPayButtonWidget =>     Padding(padding: EdgeInsets.all(10),
    child:GestureDetector(
        onTap: onTap,
        child:
        Container(
            height: 50,
            width: width,
            decoration: BoxDecoration(
              color:  Color(0xff0A6868),

              borderRadius: BorderRadius.circular(10),
              // color: HomeTabTheme.grey.withOpacity(0.2),
              boxShadow: <BoxShadow>[
                BoxShadow(
                    color:  Color(0xff0A6868),
                    offset: const Offset(1.1, 1.1),
                    blurRadius: 5.0),
              ],
            ),



            child: Image(image:AssetImage('assets/images/cashpay.png', package: 'iframe_cashpay_plugin') ,)
        )) ,);
}


lib/cash_pay_button.dart
import 'package:flutter/material.dart';

class CashPayButton extends StatelessWidget {
  const CashPayButton(
      {Key? key,
      required this.width,
      this.alignment,
      this.margin,
      this.height,
      this.padding,
      this.decoration,
      this.child,
      this.onTap})
      : super(
          key: key,
        );

  final Alignment? alignment;

  final EdgeInsetsGeometry? margin;

  final double? height;

  final double width;

  final EdgeInsetsGeometry? padding;

  final BoxDecoration? decoration;

  final Widget? child;

  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    return alignment != null
        ? Align(
            alignment: alignment ?? Alignment.center,
            child: cashPayButtonWidget,
          )
        : cashPayButtonWidget;
  }

  Widget get cashPayButtonWidget => Padding(
        padding: const EdgeInsets.all(10),
        child: GestureDetector(
            onTap: onTap,
            child: Container(
                height: 50,
                width: width,
                decoration: BoxDecoration(
                  color: const Color(0xff0A6868),

                  borderRadius: BorderRadius.circular(10),
                  // color: HomeTabTheme.grey.withOpacity(0.2),
                  boxShadow: const <BoxShadow>[
                    BoxShadow(
                        color: Color(0xff0A6868),
                        offset: Offset(1.1, 1.1),
                        blurRadius: 5.0),
                  ],
                ),
                child: const Image(
                  image: AssetImage('assets/images/cashpay.png',
                      package: 'iframe_cashpay_plugin'),
                ))),
      );
}


lib/iframe_cashpay_plugin.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_webview_cash/flutter_webview_plugin.dart';

// const kAndroidUserAgent =
//     'Mozilla/5.0 (Linux; Android 10.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Mobile Safari/537.36';

class IframeCashPay extends StatefulWidget {
  final String iframeURL;
  final JavascriptMessageHandler onConfirmPayment;
  final JavascriptMessageHandler onCancel;
  final JavascriptMessageHandler onError;
  const IframeCashPay(
      {Key? key,
      required this.iframeURL,
      required this.onConfirmPayment,
      required this.onCancel,
      required this.onError})
      : super(key: key);

  @override
  State<IframeCashPay> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<IframeCashPay> {
  final flutterWebViewPlugin = FlutterWebviewPlugin();

  // On destroy stream
  late StreamSubscription _onDestroy;

  // On urlChanged stream
  late StreamSubscription<String> _onUrlChanged;

  // On urlChanged stream
  late StreamSubscription<WebViewStateChanged> _onStateChanged;

  late StreamSubscription<WebViewHttpError> _onHttpError;

  late StreamSubscription<double> _onProgressChanged;

  late StreamSubscription<double> _onScrollYChanged;

  late StreamSubscription<double> _onScrollXChanged;

  final _history = [];

  @override
  void initState() {
    super.initState();
    flutterWebViewPlugin.close();

    // Add a listener to on destroy WebView, so you can make came actions.
    _onDestroy = flutterWebViewPlugin.onDestroy.listen((_) {
      if (mounted) {
        // Actions like show a info toast.
        ScaffoldMessenger.of(context)
            .showSnackBar(const SnackBar(content: Text('Webview Destroyed')));
      }
    });

    // Add a listener to on url changed
    _onUrlChanged = flutterWebViewPlugin.onUrlChanged.listen((String url) {
      if (mounted) {
        setState(() {
          _history.add('onUrlChanged: $url');
        });
      }
    });

    _onProgressChanged =
        flutterWebViewPlugin.onProgressChanged.listen((double progress) {
      if (mounted) {
        setState(() {
          _history.add('onProgressChanged: $progress');
        });
      }
    });

    _onScrollYChanged =
        flutterWebViewPlugin.onScrollYChanged.listen((double y) {
      if (mounted) {
        setState(() {
          _history.add('Scroll in Y Direction: $y');
        });
      }
    });

    _onScrollXChanged =
        flutterWebViewPlugin.onScrollXChanged.listen((double x) {
      if (mounted) {
        setState(() {
          _history.add('Scroll in X Direction: $x');
        });
      }
    });

    _onStateChanged =
        flutterWebViewPlugin.onStateChanged.listen((WebViewStateChanged state) {
      if (mounted) {
        setState(() {
          _history.add('onStateChanged: ${state.type} ${state.url}');
        });
      }
    });

    _onHttpError =
        flutterWebViewPlugin.onHttpError.listen((WebViewHttpError error) {
      if (mounted) {
        setState(() {
          _history.add('onHttpError: ${error.code} ${error.url}');
        });
      }
    });
  }

  onMessage(JavascriptMessage message) {
    if (message.message == "Confirmation" || message.message == "NEEDTOCHECK") {
      widget.onConfirmPayment(message);
      Navigator.pop(context);

    } else if (message.message == "Cancel") {
      widget.onCancel(message);
    } else if (message.message == "Error") {
      widget.onError(message);
    }
  }

  @override
  void dispose() {
    // Every listener should be canceled, the same should be done with this stream.
    _onDestroy.cancel();
    _onUrlChanged.cancel();
    _onStateChanged.cancel();
    _onHttpError.cancel();
    _onProgressChanged.cancel();
    _onScrollXChanged.cancel();
    _onScrollYChanged.cancel();
    flutterWebViewPlugin.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: WebviewScaffold(
        url: widget.iframeURL,
        ignoreSSLErrors: true,
        javascriptChannels: {
          JavascriptChannel(name: 'flutter', onMessageReceived: onMessage)
        },
        withLocalStorage: true,
        appCacheEnabled: false,
        resizeToAvoidBottomInset: true,
        initialChild: Container(
          color: Colors.white,
          child: const Center(
            child: Text('Waiting.....'),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:iframe_cashpay_plugin/iframe_cashpay_plugin.dart';
import 'package:iframe_cashpay_plugin/cash_pay_button.dart';

void main() {
  runApp(const PayMaterialApp());
}

class PayMaterialApp extends StatelessWidget {
  const PayMaterialApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Pay for Flutter Demo',
      home: PaySampleApp(title: 'PaySampleApp'),
    );
  }
}

class PaySampleApp extends StatefulWidget {
  final String title;
  const PaySampleApp({Key? key, required this.title}) : super(key: key);

  @override
  State<PaySampleApp> createState() => PaySampleAppState();
}

class PaySampleAppState extends State<PaySampleApp> {
  String orderId = "";

  /// This function Send itemList for yor server.
  ///
  /// @param itemList.
  /// @return iframeURL and orderId.
  ///
  /// Example:
  ///
  /// ```
  /// var res = sendItems(itemList);
  /// print(res.then((iframeURL)=>iframeURL)); // "https://########"
  /// ```
  Future<String> sendItems(itemList) async {
    //Send itemList for yor server and post CreateOrder.
    //iframeURL returned from Response CreateOrder
    //Documentation https://documenter.getpostman.com/view/17550185/2s93XzwN9o
    String iframeURL = "";
    //orderID returned from Response CreateOrder
    //Store the orderID in the orderId variable to use on function onConfirmPayment
    orderId = "";
    return iframeURL;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        backgroundColor: Colors.white,
        body: ListView(children: [
          CashPayButton(
              width: MediaQuery.of(context).size.width,
              onTap: () async {
                await sendItems({
                  {
                    "itemName": "كتاب",
                    "amount": 2000,
                  },
                  {
                    "itemName": "ساعة",
                    "amount": 5000,
                  }
                }).then((iframeURL) => showModalBottomSheet(
                    context: context,
                    isDismissible: false,
                    shape: const RoundedRectangleBorder(
                      borderRadius: BorderRadius.vertical(
                        top: Radius.circular(30.0),
                      ),
                    ),
                    isScrollControlled: true,
                    builder: (BuildContext context) {
                      return Container(
                          padding: const EdgeInsets.only(
                              left: 10.0, right: 10.0, top: 10.0),
                          height: MediaQuery.of(context).size.height * 0.7,
                          //IframeCashPay widget displays the Cash E-wallet payment iframe.
                          /// This widget IframeCashPay displays the Cash E-wallet payment iframe.
                          ///
                          /// @param iframeURL .
                          /// @param onConfirmPayment.
                          /// @param onCancel.
                          /// @param onError.
                          /// @return message onConfirmPayment or onCancel or onError.
                          ///
                          /// Example:
                          ///
                          /// ```
                          ///  IframeCashPay( iframeURL: "https://########",
                          ///    onConfirmPayment: onConfirmPayment,
                          ///    onCancel: onCancel,
                          ///    onError: onError,);
                          /// ```
                          child: IframeCashPay(
                            iframeURL: iframeURL,
                            onConfirmPayment: onConfirmPayment,
                            onCancel: onCancel,
                            onError: onError,
                          ));
                    }));
              })
        ]));
  }

  /// This function callback onConfirmPayment.
  ///
  /// @param message.
  /// @return message onConfirmPayment.
  ///
  /// Example:
  ///
  /// ```
  /// onConfirmPayment("NEEDTOCHECK");
  /// ```
  onConfirmPayment(message) {
    //After Confirmatin from iFrameCashPay.
    if (orderId.isNotEmpty) {
      //Here use CheckOrderStatus on your server to check order status.
      //Documentation https://documenter.getpostman.com/view/17550185/2s93XzwN9o
    }

  }

  /// This function callback onCancel.
  ///
  /// @param message.
  /// @return message onCancel.
  ///
  /// Example:
  ///
  /// ```
  /// onCancel("Cancel");
  /// ```
  onCancel(message) {
    //After Cancel from iFrameCashPay.
    Navigator.pop(context);
  }

  /// This function callback onError.
  ///
  /// @param message.
  /// @return message onError.
  ///
  /// Example:
  ///
  /// ```
  /// onError("Error");
  /// ```
  onError(message) {
    //After return Error from iFrameCashPay.
    Navigator.pop(context);
  }
}
